### PR TITLE
[Trigger] refactor Converter

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This will ease versioning of trigger converter we will need to introduce interface support.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala [hunk 1]
```diff
@@ -34,205 +36,133 @@ import com.daml.platform.participant.util.LfEngineToApi.{
...
-case class Converter(
-    fromTransaction: Transaction => Either[String, SValue],
-    fromCompletion: Completion => Either[String, SValue],
-    fromHeartbeat: SValue,
-    fromACS: Seq[CreatedEvent] => Either[String, SValue],
-    toFiniteDuration: SValue => Either[String, FiniteDuration],
-    toCommands: SValue => Either[String, Seq[Command]],
-    toRegisteredTemplates: SValue => Either[String, Seq[Identifier]],
-)
+final class Converter(compiledPackages: CompiledPackages, triggerIds: TriggerIds) {
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala:39]: I'd always thought this was an odd way to define the converter - use a case class and instantiate via an apply!

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala [hunk 2]
```diff
@@ -34,205 +36,133 @@ import com.daml.platform.participant.util.LfEngineToApi.{
...
-    final case class Template(name: ChoiceName, arg: SValue) extends AnyChoice
-    final case class Interface(ifaceId: Identifier, name: ChoiceName, arg: SValue) extends AnyChoice
-  }
-  private case class AnyContractKey(key: SValue)
+  private[this] val valueTranslator = new preprocessing.ValueTranslator(
+    compiledPackages.pkgInterface,
+    requireV1ContractIdSuffix = false,
+  )
+
+  private[this] def validateRecord(r: value.Record): Either[String, Value.ValueRecord] =
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala:49]: Given the method (here and elsewhere) has no state, surely its safe to share this method (i.e. not use the `this` restriction)? Simply thinking its better to share the method code per class than per object.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala [hunk 3]
```diff
@@ -34,205 +36,133 @@ import com.daml.platform.participant.util.LfEngineToApi.{
...
+    NoLoggingValueValidator.validateRecord(r).left.map(_.getMessage)
+
+  private[this] def translateValue(ty: Type, value: Value): Either[String, SValue] =
+    valueTranslator.translateValue(ty, value).left.map(res => s"Failure to translate value: $res")
+
+  private[this] val anyTemplateTyCon = DA.Internal.Any.assertIdentifier("AnyTemplate")
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala:55]: Likewise, for this set of (type/Ty) val's, maybe the `this` restriction isn't needed?

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala [hunk 4]
```diff
@@ -525,33 +447,73 @@ object Converter {
...
+    val SucceedVariantConstrcutor = 1
   }
+}
 
-  def apply(compiledPackages: CompiledPackages, triggerIds: TriggerIds): Converter = {
-    val valueTranslator = new preprocessing.ValueTranslator(
-      compiledPackages.pkgInterface,
-      requireV1ContractIdSuffix = false,
+// Helper to create identifiers pointing to the Daml.Trigger module
+case class TriggerIds(val triggerPackageId: PackageId) {
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala:497]: Should we make this final?

**Review by carlpulley-da — APPROVED** [review]
Some small niggles, but ow LGTM. Nicely sweeps up some code sores!

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala:49]: I prefer to use the same annotation consistently, it becomes very quickly tricky to reason about which method has state or not. 
I favored `private[this]` over `private` are it is more restrictive. I would probably be fine to change my default style  and use `private` everywhere, but let's discuss that another time.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Converter.scala:497]: :+1:

**Pull request merged**